chemclipse/plugins/org.eclipse.chemclipse.msd.model/src/org/eclipse/chemclipse/msd/model/matrix/ExtractedMatrix.java

```diff
@@ -0,0 +1,189 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.chemclipse.msd.model.core.IIon;
+import org.eclipse.chemclipse.msd.model.core.IIonBounds;
+import org.eclipse.chemclipse.msd.model.core.IScanMSD;
+import org.eclipse.chemclipse.msd.model.core.selection.IChromatogramSelectionMSD;
+import org.eclipse.chemclipse.msd.model.implementation.Ion;
+
+public class ExtractedMatrix implements IExtractedMatrix {
+	
+	private IChromatogramSelectionMSD selection;
+	private List<IScanMSD> scans;
+	private int numberOfScans;
+	private int numberOfIons;
+	private int startIon;
+	private int stopIon;
+	private float[] signal;
+	private boolean highRes = false;
+	
+	public ExtractedMatrix(IChromatogramSelectionMSD chromatogramSelection) {
+		this.selection = chromatogramSelection;
+		this.numberOfScans = selection.getStopScan() - selection.getStartScan() - 1;
+		this.scans = extractScans();
+		if(checkHighRes( 10)) {
+			highRes = true;
+			System.out.println("This is highres MS Data and should be binned first");
+		} else {
+			this.startIon = getMinMz();
+			this.stopIon = getMaxMz();
+			this.numberOfIons = this.stopIon - this.startIon + 1;
+			signal = new float[numberOfScans * (this.stopIon - this.startIon + 1)];
+			try {
+				for(int i = 0; i < numberOfScans; i++) {
+					for(int j = 0; j < scans.get(i).getIons().size(); j++) {
+						signal[ ((int) Math.round(scans.get(i).getIons().get(j).getIon() - this.startIon)) * numberOfScans + i    ] =  scans.get(i).getIons().get(j).getAbundance();
+					}
+				}
+			} catch (Exception e) {
+				e.printStackTrace();
+			}
+		}
+	}
+	
+	private Boolean checkHighRes( int limit ) {
+		boolean moreIonsThanFullMz = false;
+		Iterator<IScanMSD> iterator = scans.iterator();
+		IIonBounds bounds;
+		double rangeAbs;
+		IScanMSD currentScan;
+		
+		while(moreIonsThanFullMz == false) {
+			currentScan = iterator.next();
+			bounds = currentScan.getIonBounds();
+			rangeAbs = bounds.getHighestIon().getIon() - bounds.getLowestIon().getIon();
+			if(rangeAbs + limit < currentScan.getIons().size()) {
+				moreIonsThanFullMz = true;
+			}
+			if(!iterator.hasNext()) {
+				break;
+			}
+		}
+		
+		return moreIonsThanFullMz;
+	}
+	
+	private List<IScanMSD> extractScans() {
+
+		List<IScanMSD> scans;
+		int startRT;
+		int stopRT;
+		startRT = selection.getStartRetentionTime();
+		stopRT = selection.getStopRetentionTime();
+		scans = selection.getChromatogram() //
+				.getScans() //
+				.stream() //
+				.filter(s -> s instanceof IScanMSD) //
+				.map(IScanMSD.class::cast) //
+				.filter(s -> s.getRetentionTime() > startRT) //
+				.filter(s -> s.getRetentionTime() < stopRT) //
+				.collect(Collectors.toList()); //
+		return (scans);
+	}
+	
+	private int getMinMz() {
+
+		double min = scans.stream() //
+				.flatMap(scan -> scan.getIons().stream()) //
+				.min(Comparator.comparing(IIon::getIon)) //
+				.get() //
+				.getIon();
+		int minMz = (int)Math.round(min);
```

Comment:
You can use org.eclipse.chemclipse.msd.model.core.AbstractIon.getIon(double) instead, even though its functional equivalent its more descriptive and allows to change rounding sceme in a central place.